File: admin/src/pilotReadiness.js
const CRITICAL_CHECKS = Object.freeze([
  "database",
  "migrations",
  "env",
  "payments",
  "moysklad",
  "scheduler",
  "notification_delivery",
  "webhook_outbox",
  "moysklad_sync",
]);
const ADVISORY_CHECKS = Object.freeze(["media", "search"]);
const PILOT_STATUSES = new Set(["not_armed", "active", "stopped", "completed"]);

const CHECK_LABELS = Object.freeze({
  database: "База данных",
  migrations: "Миграции БД",
  env: "Production-конфигурация",
  payments: "YooKassa",
  moysklad: "MoySklad",
  scheduler: "Scheduler",
  notification_delivery: "Доставка Telegram-уведомлений",
  webhook_outbox: "Webhook outbox",
  moysklad_sync: "Синхронизация MoySklad",
  media: "Media storage",
  search: "Поиск",
});

const READINESS_LABELS = Object.freeze({
  diagnostics_unavailable: "Service diagnostics недоступны",
  runtime_status_unavailable: "Pilot runtime status недоступен",
  runtime_checkout_no_go: "Pilot runtime запретил следующий checkout",
  runtime_database_integrity_failed: "Нарушена целостность runtime БД",
  runtime_artifact_integrity_unavailable: "Runtime/release evidence недоступен",
  runtime_artifact_integrity_failed: "Runtime/release evidence не прошёл проверку",
  runtime_money_attention: "Денежная операция требует проверки",
  runtime_operational_safety_unavailable: "Проверка operational queues недоступна",
  runtime_operational_safety_failed: "Operational queues требуют вмешательства",
  response_contract_invalid: "Ответ readiness имеет неверный или небезопасный формат",
});

function isObject(value) {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function isSafeInteger(value) {
  return Number.isSafeInteger(value) && value >= 0;
}

function isSafeIntegerOrNull(value) {
  return value === null || isSafeInteger(value);
}

function isBooleanOrNull(value) {
  return value === null || typeof value === "boolean";
}

function diagnosticCodeParts(code) {
  if (typeof code !== "string") return null;
  const match = /^diagnostic_(failed|missing|degraded):([a-z_]+)$/.exec(code);
  if (!match) return null;
  return { state: match[1], check: match[2] };
}

function isAllowedBlockingCode(code) {
  if (typeof code !== "string") return false;
  if (Object.hasOwn(READINESS_LABELS, code) && code !== "response_contract_invalid") return true;
  const parts = diagnosticCodeParts(code);
  return Boolean(
    parts
    && (parts.state === "failed" || parts.state === "missing")
    && CRITICAL_CHECKS.includes(parts.check),
  );
}

function isAllowedWarningCode(code) {
  const parts = diagnosticCodeParts(code);
  return Boolean(
    parts
    && (parts.state === "missing" || parts.state === "degraded")
    && ADVISORY_CHECKS.includes(parts.check),
  );
}

function normalizeCodes(value, validator) {
  if (!Array.isArray(value)) {
    return { codes: ["response_contract_invalid"], valid: false };
  }
  const codes = [];
  let valid = true;
  for (const code of value) {
    if (validator(code)) codes.push(code);
    else valid = false;
  }
  if (!valid) codes.push("response_contract_invalid");
  return { codes: [...new Set(codes)], valid };
}

function normalizeDiagnosticGroup(value, expected) {
  if (!isObject(value)) return { values: {}, valid: false };
  const values = {};
  let valid = true;
  for (const name of expected) {
    const item = value[name];
    if (typeof item === "boolean" || item === null) values[name] = item;
    else {
      values[name] = null;
      valid = false;
    }
  }
  return { values, valid };
}

function fallbackReadiness() {
  return {
    decision: "NO-GO",
    readyForNextOrder: false,
    contractValid: false,
    blockingCodes: ["response_contract_invalid"],
    warningCodes: [],
    critical: Object.fromEntries(CRITICAL_CHECKS.map((name) => [name, null])),
    advisory: Object.fromEntries(ADVISORY_CHECKS.map((name) => [name, null])),
  };
}

export function normalizePilotReadiness(payload) {
  if (!isObject(payload) || payload.schema_version !== 1) return fallbackReadiness();

  const blocking = normalizeCodes(payload.blocking_codes, isAllowedBlockingCode);
  const warnings = normalizeCodes(payload.warning_codes, isAllowedWarningCode);
  const diagnostics = isObject(payload.diagnostics) ? payload.diagnostics : {};
  const critical = normalizeDiagnosticGroup(diagnostics.critical, CRITICAL_CHECKS);
  const advisory = normalizeDiagnosticGroup(diagnostics.advisory, ADVISORY_CHECKS);
  const runtime = isObject(payload.runtime) ? payload.runtime : {};

  const decisionValid = payload.decision === "GO" || payload.decision === "NO-GO";
  const readyValid = typeof payload.ready_for_next_order === "boolean";
  const runtimeValid = (
    (runtime.checkout_decision === "GO" || runtime.checkout_decision === "NO-GO")
    && typeof runtime.enforced === "boolean"
    && (runtime.status === null || (typeof runtime.status === "string" && PILOT_STATUSES.has(runtime.status)))
    && isSafeIntegerOrNull(runtime.accepted_orders)
    && isSafeIntegerOrNull(runtime.remaining_orders)
    && isSafeIntegerOrNull(runtime.allowlist_count)
    && isBooleanOrNull(runtime.database_integrity_healthy)
    && isBooleanOrNull(runtime.artifact_integrity_applicable)
    && isBooleanOrNull(runtime.artifact_integrity_healthy)
    && typeof runtime.money_attention_required === "boolean"
    && isBooleanOrNull(runtime.operational_safety_applicable)
    && isBooleanOrNull(runtime.operational_safety_healthy)
  );
  const decisionCoherent = decisionValid
    && readyValid
    && ((payload.decision === "GO") === payload.ready_for_next_order);
  const goSignalsConfirmed = payload.decision !== "GO" || (
    blocking.codes.length === 0
    && CRITICAL_CHECKS.every((name) => critical.values[name] === true)
    && runtime.checkout_decision === "GO"
    && runtime.enforced === true
    && runtime.status === "active"
    && isSafeInteger(runtime.remaining_orders)
    && runtime.remaining_orders > 0
    && runtime.database_integrity_healthy === true
    && runtime.artifact_integrity_applicable === true
    && runtime.artifact_integrity_healthy === true
    && runtime.money_attention_required === false
    && runtime.operational_safety_applicable === true
    && runtime.operational_safety_healthy === true
  );
  const contractValid = Boolean(
    blocking.valid
    && warnings.valid
    && critical.valid
    && advisory.valid
    && runtimeValid
    && decisionCoherent
    && goSignalsConfirmed
  );

  const blockingCodes = contractValid
    ? blocking.codes
    : [...new Set([...blocking.codes, "response_contract_invalid"])];
  const decision = contractValid
    && payload.decision === "GO"
    && payload.ready_for_next_order === true
    && blockingCodes.length === 0
    ? "GO"
    : "NO-GO";

  return {
    decision,
    readyForNextOrder: decision === "GO",
    contractValid,
    blockingCodes,
    warningCodes: warnings.codes.filter((code) => code !== "response_contract_invalid"),
    critical: critical.values,
    advisory: advisory.values,
  };
}

export function pilotReadinessCodeLabel(code) {
  if (Object.hasOwn(READINESS_LABELS, code)) return READINESS_LABELS[code];
  const parts = diagnosticCodeParts(code);
  if (!parts || !Object.hasOwn(CHECK_LABELS, parts.check)) {
    return READINESS_LABELS.response_contract_invalid;
  }
  const check = CHECK_LABELS[parts.check];
  if (parts.state === "failed") return `${check}: проверка не пройдена`;
  if (parts.state === "missing") return `${check}: сигнал отсутствует`;
  if (parts.state === "degraded") return `${check}: деградация`;
  return READINESS_LABELS.response_contract_invalid;
}

export function pilotDiagnosticLabel(name) {
  return CHECK_LABELS[name] || "Неизвестная проверка";
}

File: backend/services/pilot_readiness.py
from __future__ import annotations

from typing import Any, Mapping, TYPE_CHECKING

from sqlalchemy.orm import Session

from .diagnostics import run_diagnostics
from .pilot_observability import build_pilot_operations_status

if TYPE_CHECKING:
    from ..config import Settings

_CRITICAL_DIAGNOSTIC_CHECKS = (
    "database",
    "migrations",
    "env",
    "payments",
    "moysklad",
    "scheduler",
    "notification_delivery",
    "webhook_outbox",
    "moysklad_sync",
)
_ADVISORY_DIAGNOSTIC_CHECKS = ("media", "search")


def _diagnostic_status(snapshot: Mapping[str, Any], name: str) -> bool | None:
    checks = snapshot.get("checks")
    if not isinstance(checks, Mapping):
        return None
    check = checks.get(name)
    if not isinstance(check, Mapping):
        return None
    value = check.get("ok")
    return value if isinstance(value, bool) else None


def _runtime_summary(snapshot: Mapping[str, Any]) -> dict[str, Any]:
    runtime = snapshot.get("runtime") if isinstance(snapshot.get("runtime"), Mapping) else {}
    database = snapshot.get("database_integrity") if isinstance(snapshot.get("database_integrity"), Mapping) else {}
    artifact = snapshot.get("artifact_integrity") if isinstance(snapshot.get("artifact_integrity"), Mapping) else {}
    continuation = snapshot.get("continuation") if isinstance(snapshot.get("continuation"), Mapping) else {}
    money = snapshot.get("money_attention") if isinstance(snapshot.get("money_attention"), Mapping) else {}
    operational = snapshot.get("operational_safety") if isinstance(snapshot.get("operational_safety"), Mapping) else {}
    return {
        "checkout_decision": str(snapshot.get("checkout_decision") or "NO-GO"),
        "enforced": bool(snapshot.get("enforced")),
        "status": runtime.get("status"),
        "accepted_orders": runtime.get("accepted_orders"),
        "remaining_orders": runtime.get("remaining_orders"),
        "allowlist_count": runtime.get("allowlist_count"),
        "database_integrity_healthy": database.get("healthy"),
        "artifact_integrity_applicable": artifact.get("applicable"),
        "artifact_integrity_healthy": artifact.get("healthy"),
        "continuation_applicable": continuation.get("applicable"),
        "continuation_ready": continuation.get("ready"),
        "next_sequence": continuation.get("next_sequence"),
        "money_attention_required": bool(money.get("attention_required")),
        "operational_safety_applicable": operational.get("applicable"),
        "operational_safety_healthy": operational.get("healthy"),
    }


def compose_pilot_readiness(
    diagnostics: Mapping[str, Any] | None,
    runtime_status: Mapping[str, Any] | None,
) -> dict[str, Any]:
    blocking_codes: list[str] = []
    warning_codes: list[str] = []

    diagnostics_available = isinstance(diagnostics, Mapping) and bool(diagnostics)
    runtime_available = isinstance(runtime_status, Mapping) and bool(runtime_status)
    diagnostics = diagnostics if diagnostics_available else {}
    runtime_status = runtime_status if runtime_available else {}

    if not diagnostics_available:
        blocking_codes.append("diagnostics_unavailable")
    else:
        for name in _CRITICAL_DIAGNOSTIC_CHECKS:
            status = _diagnostic_status(diagnostics, name)
            if status is None:
                blocking_codes.append(f"diagnostic_missing:{name}")
            elif status is False:
                blocking_codes.append(f"diagnostic_failed:{name}")
        for name in _ADVISORY_DIAGNOSTIC_CHECKS:
            status = _diagnostic_status(diagnostics, name)
            if status is None:
                warning_codes.append(f"diagnostic_missing:{name}")
            elif status is False:
                warning_codes.append(f"diagnostic_degraded:{name}")

    if not runtime_available:
        blocking_codes.append("runtime_status_unavailable")
    elif runtime_status.get("checkout_decision") != "GO":
        blocking_codes.append("runtime_checkout_no_go")

    runtime = _runtime_summary(runtime_status)
    if runtime_available:
        if runtime.get("database_integrity_healthy") is not True:
            blocking_codes.append("runtime_database_integrity_failed")
        if runtime.get("artifact_integrity_applicable") is not True:
            blocking_codes.append("runtime_artifact_integrity_unavailable")
        elif runtime.get("artifact_integrity_healthy") is not True:
            blocking_codes.append("runtime_artifact_integrity_failed")
        if runtime.get("continuation_applicable") is True and runtime.get("continuation_ready") is False:
            blocking_codes.append("runtime_previous_scenario_pending")
        if runtime.get("money_attention_required"):
            blocking_codes.append("runtime_money_attention")
        if runtime.get("operational_safety_applicable") is not True:
            blocking_codes.append("runtime_operational_safety_unavailable")
        elif runtime.get("operational_safety_healthy") is not True:
            blocking_codes.append("runtime_operational_safety_failed")

    blocking_codes = sorted(set(blocking_codes))
    warning_codes = sorted(set(warning_codes))
    ready = not blocking_codes and runtime.get("checkout_decision") == "GO"

    critical_checks = {
        name: _diagnostic_status(diagnostics, name) if diagnostics_available else None
        for name in _CRITICAL_DIAGNOSTIC_CHECKS
    }
    advisory_checks = {
        name: _diagnostic_status(diagnostics, name) if diagnostics_available else None
        for name in _ADVISORY_DIAGNOSTIC_CHECKS
    }

    return {
        "schema_version": 1,
        "decision": "GO" if ready else "NO-GO",
        "ready_for_next_order": ready,
        "blocking_codes": blocking_codes,
        "warning_codes": warning_codes,
        "diagnostics": {
            "critical": critical_checks,
            "advisory": advisory_checks,
        },
        "runtime": runtime,
    }


def build_pilot_readiness(
    db: Session,
    settings: "Settings",
    *,
    env: Mapping[str, str] | None = None,
) -> dict[str, Any]:
    diagnostics: Mapping[str, Any] | None
    runtime_status: Mapping[str, Any] | None

    try:
        diagnostics = run_diagnostics(db)
    except Exception:
        diagnostics = None

    try:
        runtime_status = build_pilot_operations_status(db, settings, env=env)
    except Exception:
        runtime_status = None

    return compose_pilot_readiness(diagnostics, runtime_status)


__all__ = ["build_pilot_readiness", "compose_pilot_readiness"]